src/src_cryptography/src_libgcrypt/cryptography_interface_libgcrypt.template.c

```diff
@@ -0,0 +1,778 @@
+
+#include "crypto.h"
+#include "crypto_error.h"
+#include "cryptography_interface.h"
+
+
+// Cryptography Interface Initialization & Management Functions
+static int32_t cryptography_config(void);
+static int32_t cryptography_init(void);
+static crypto_key_t* get_ek_ring(void);
+static int32_t cryptography_shutdown(void);
+// Cryptography Interface Functions
+static int32_t cryptography_encrypt(void);
+static int32_t cryptography_decrypt(void);
+static int32_t cryptography_authenticate(void);
+static int32_t cryptography_validate_authentication(void);
+static int32_t cryptography_aead_encrypt(uint8_t* data_out, uint32_t len_data_out,
+                                         uint8_t* data_in, uint32_t len_data_in,
+                                         uint8_t* key, uint32_t len_key,
+                                         SecurityAssociation_t* sa_ptr,
+                                         uint8_t* iv, uint32_t iv_len,
+                                         uint8_t* mac, uint32_t mac_size,
+                                         uint8_t* aad, uint32_t aad_len,
+                                         uint8_t encrypt_bool, uint8_t authenticate_bool,
+                                         uint8_t aad_bool);
+static int32_t cryptography_aead_decrypt(uint8_t* data_out, uint32_t len_data_out,
+                                         uint8_t* data_in, uint32_t len_data_in,
+                                         uint8_t* key, uint32_t len_key,
+                                         SecurityAssociation_t* sa_ptr,
+                                         uint8_t* iv, uint32_t iv_len,
+                                         uint8_t* mac, uint32_t mac_size,
+                                         uint8_t* aad, uint32_t aad_len,
+                                         uint8_t decrypt_bool, uint8_t authenticate_bool,
+                                         uint8_t aad_bool);
+/*
+** Module Variables
+*/
+// Security
+static crypto_key_t ek_ring[NUM_KEYS] = {0};
+// Cryptography Interface
+static CryptographyInterfaceStruct cryptography_if_struct;
+
+CryptographyInterface get_cryptography_interface_libgcrypt(void)
+{
+    cryptography_if_struct.cryptography_config = cryptography_config;
+    cryptography_if_struct.cryptography_init = cryptography_init;
+    cryptography_if_struct.get_ek_ring = get_ek_ring;
+    cryptography_if_struct.cryptography_shutdown = cryptography_shutdown;
+    cryptography_if_struct.cryptography_encrypt = cryptography_encrypt;
+    cryptography_if_struct.cryptography_decrypt = cryptography_decrypt;
+    cryptography_if_struct.cryptography_authenticate = cryptography_authenticate;
+    cryptography_if_struct.cryptography_validate_authentication = cryptography_validate_authentication;
+    cryptography_if_struct.cryptography_aead_encrypt = cryptography_aead_encrypt;
+    cryptography_if_struct.cryptography_aead_decrypt = cryptography_aead_decrypt;
+    return &cryptography_if_struct;
```

Comment:
Should we distinguish these function names between the different types?  eg kmc_cryptography_config

Reply:
We could do that, but I think that would actually make working with the code base somewhat trickier... If we ever need to update the interface, we generally need to update it in all the places where the functions are defined -- if they have different names, search and replace won't work, and you'll need to manually map out what goes to what in your IDE.

Having a unique name here doesn't add much, just knowing that the functions are static (IE locally defined in this code file) should be enough of an indication about what's going on.